modules/dbfs/couchbase.go

```diff
@@ -168,74 +167,162 @@ func (di *DatabaseImpl) CBAppendFileChange(fileID int64, changes, prevChanges []
 	// then use it in the MutateIn call to verify the document hasn't updated underneath us
 	frag, err := cb.bucket.LookupIn(key).Get("version").Get("usetemp").Execute()
 	if err != nil {
-		return -1, nil, ErrResourceNotFound
+		return nil, -1, nil, ErrResourceNotFound
 	}
 
 	cas := frag.Cas()
 	var version int64
 	var useTemp bool
 	err = frag.Content("version", &version)
 	if err != nil {
-		return -1, nil, ErrNoData
+		return nil, -1, nil, ErrNoData
 	}
 	err = frag.Content("usetemp", &useTemp)
 	if err != nil {
-		return -1, nil, ErrNoData
+		return nil, -1, nil, ErrNoData
 	}
 
-	minVersion := int64(math.MaxInt64)
-	transformedChanges := make([]string, len(changes))
+	minStartIndex := int64(math.MaxInt64)
+	transformedPatches := []string{}
 
 	// Build patch, transform changes against newer changes.
-	for i, changeStr := range changes {
+	for _, changeStr := range patches {
 		change, err := patching.NewPatchFromString(changeStr)
 		if err != nil {
-			return -1, nil, errors.New("Failed to parse patch")
+			return nil, -1, nil, errors.New("Failed to parse patch")
 		}
 
 		// check to make sure the patch is being applied to the most recent revision
 		if change.BaseVersion > version {
-			return -1, nil, ErrVersionOutOfDate
+			utils.LogError("BaseVersion too high", ErrVersionOutOfDate, nil)
+			return nil, -1, nil, ErrVersionOutOfDate
 		}
 
-		if minVersion > change.BaseVersion {
-			minVersion = change.BaseVersion
-		}
 		// For every patch, calculate the patches that it does not have.
-		startIndex := len(prevChanges) - int(version-change.BaseVersion)
+		utils.LogDebug("CHANGES VERSIONS", utils.LogFields{
+			"Version":     version,
+			"BaseVersion": change.BaseVersion,
+			"Diff":        int(version - change.BaseVersion),
+			"Len":         len(prevChanges),
+			"ChangeStr":   changeStr,
+			"PrevChanges": prevChanges,
+		})
+
+		//startIndex := len(prevChanges) - int(version-change.BaseVersion)
+		//if startIndex < 0 {
+		//	utils.LogError("StartIndex is negative", ErrVersionOutOfDate, nil)
+		//	return nil, -1, nil, ErrVersionOutOfDate
+		//}
+
+		startIndex := int64(len(prevChanges) - 1)
+
+		// If we are building on the server's base version, don't need to transform.
+		if change.BaseVersion == version {
+			startIndex = int64(len(prevChanges))
+		} else {
+			for startIndex >= 0 && startIndex < int64(len(prevChanges)) {
+				otherPatch, err := patching.NewPatchFromString(prevChanges[startIndex])
+
+				utils.LogDebug("CHECKING", utils.LogFields{
+					"Change":     changeStr,
+					"OtherPatch": prevChanges[startIndex],
+					"Result":     change.BaseVersion > otherPatch.BaseVersion,
+					"StartIndex": startIndex,
+				})
+
+				if err != nil {
+					return nil, -1, nil, ErrInternalServerError
```

Comment:
TBH, this should never happen. Every patch is parsed before going into the database, so all patches should be able to be parsed coming out of it.